File: src/lib/offlineAuth.ts
/**
 * #666 — offline-first identity on the web: what a failed `GET /api/auth/me`
 * means for routing and for the local state the app already holds.
 *
 * The TypeScript mirror of the iOS `OfflineAuth` landed in #665 (PR #696), and
 * deliberately the *same* three moving parts rather than a second convention:
 *
 *   1. one classifier that maps a failed `me()` onto a sign-out taxonomy,
 *   2. one shared predicate (`mayClearAccountScopedState`) deciding whether a
 *      cause is authoritative enough to destroy local state — used by every
 *      consumer (routing, the cached identity, the tracking-unlock reset),
 *   3. one outcome that keeps the user signed in from a cached identity
 *      whenever that predicate says no and a cached identity exists.
 *
 * The practice needs nothing from the server: starting a sit, watching the
 * timer, tapping thoughts, finishing. Internet is required for exactly one
 * thing — *syncing*. Knowing who you are is not syncing, so a request that
 * never reached the server must not be read as a sign-out. Before this, any
 * failed `me()` dropped the installed PWA on an auth error, which is precisely
 * what happens on a plane, on the subway, or in a dead zone.
 *
 * The asymmetry is deliberate. Wrongly deciding "offline" costs a 401 on the
 * next request and a trip to sign-in then; wrongly deciding "signed out"
 * destroys local state and is the bug this exists to prevent. Only an
 * authoritative signal gets to be destructive.
 *
 * **Two deliberate deviations from a literal port of the Swift classifier**,
 * both because the web surface produces answers the iOS `APIError` surface does
 * not, and both preserving the pre-#666 web behavior rather than inventing new:
 *
 * - `403` joins `401` as authoritative. iOS only ever sees `401` for a rejected
 *   session; the web bootstrap has always treated both as a rejection, and the
 *   #666 acceptance criteria name the pair.
 * - `404` is authoritative as `signedOut`. It is this route's "the session was
 *   valid but the account row is gone" answer (`/api/auth/me` returns
 *   `404 User not found` after a deletion) — the analogue of the iOS
 *   `applySignedOut(cause: .signedOut)` nil-user branch, not of a transport
 *   failure.
 *
 * Everything else the server said, and every answer that never arrived, stays
 * non-authoritative exactly as on iOS.
 */

/**
 * Why the app believes it may not have a session. Mirrors the iOS
 * `WidgetDataStore.SignedOutCause` taxonomy #676 introduced for the widget, so
 * the two clients can never disagree about what "signed out" means.
 */
export type SignedOutCause =
  /** The user signed out, or the server authoritatively reported no account. */
  | "signedOut"
  /** Credentials were rejected (HTTP 401/403) — the session is genuinely over. */
  | "unauthorized"
  /** The server answered, but with an error that says nothing about auth. */
  | "serverError"
  /** The server could not be reached at all (offline cold start). */
  | "unreachable";

/** A `me()` attempt that did not yield an authenticated user. */
export type MeFailure =
  /** The server answered with a non-2xx status. */
  | { kind: "status"; status: number }
  /** `fetch` rejected: no response ever arrived (offline, DNS, timeout, abort). */
  | { kind: "transport" };

/** What the app should do after a failed attempt to load the current user. */
export type AuthBootstrapOutcome = {
  /**
   * - `offline` — stay signed in and render from the cached identity.
   * - `signedOut` — the session is provably over; render sign-in.
   * - `unavailable` — no trustworthy answer *and* nothing cached to render;
   *   render the retryable error. This is the web's rendering of the iOS
   *   `.signedOut(cause)` "nothing to fall back to" branch: like it, it clears
   *   nothing, because the cause was never authoritative.
   */
  action: "offline" | "signedOut" | "unavailable";
  cause: SignedOutCause;
  /**
   * The single answer to "may this cause destroy account-scoped local state?",
   * resolved once here so routing, the cached identity and the tracking-unlock
   * reset cannot drift apart by each asking separately.
   */
  clearsLocalState: boolean;
  /** Whether the app is about to run from its local copy. Drives the indicator. */
  usesCachedIdentity: boolean;
};

/**
 * Map a failed `me()` onto the shared sign-out taxonomy.
 *
 * Only an answer that came *from the server* and was *about auth* counts as
 * proof the session is over. A response-less failure is `unreachable`; a status
 * the transport never produced (`<= 0`, how some clients model a dead network)
 * is treated the same way, mirroring the Swift `status <= 0` branch.
 */
export function signedOutCauseFor(failure: MeFailure): SignedOutCause {
  if (failure.kind === "transport") return "unreachable";
  const { status } = failure;
  if (status === 401 || status === 403) return "unauthorized";
  if (status === 404) return "signedOut";
  if (status <= 0) return "unreachable";
  return "serverError";
}

/**
 * The one predicate for "is this cause authoritative enough to destroy local
 * state?" — the direct mirror of `WidgetDataStore.shouldClearStoredSnapshot`.
 * A dropped connection or a 5xx is never one of them.
 */
export function mayClearAccountScopedState(cause: SignedOutCause): boolean {
  return cause === "signedOut" || cause === "unauthorized";
}

/**
 * Resolve a failed `me()` into a route.
 *
 * `hasCachedUser` is the caller's answer for `loadCachedUser() !== null`;
 * without a local copy there is nothing to render, so even a transport failure
 * has to land somewhere else (a first launch with no network, say).
 */
export function resolveAuthBootstrap(
  failure: MeFailure,
  hasCachedUser: boolean,
): AuthBootstrapOutcome {
  const cause = signedOutCauseFor(failure);
  const clearsLocalState = mayClearAccountScopedState(cause);

  if (clearsLocalState || !hasCachedUser) {
    return {
      action: clearsLocalState ? "signedOut" : "unavailable",
      cause,
      clearsLocalState,
      usesCachedIdentity: false,
    };
  }

  return { action: "offline", cause, clearsLocalState: false, usesCachedIdentity: true };
}

/**
 * Status line for a route away from the app. Unchanged from the pre-#666
 * copy — a server problem and a dead network read differently to the user, and
 * both are retryable.
 */
export function authErrorMessageFor(cause: SignedOutCause): string {
  return cause === "unreachable"
    ? "Unable to verify your sign-in due to a network issue. Please retry."
    : "Unable to verify your sign-in due to a server issue. Please retry.";
}

File: ios/StillPointShared/Sources/StillPointShared/OfflineAuth.swift
import Foundation

/// #665 — offline-first identity: what a failed `/api/auth/me` means for routing
/// and for the local state the app already holds.
///
/// The practice itself needs nothing from the server: starting a sit, watching
/// the timer, tapping thoughts, finishing. Internet is required for exactly one
/// thing — *syncing*. Knowing who you are is not syncing, so a request that never
/// reached the server must not be read as a sign-out. Before this, any failed
/// `me()` dropped the user on the sign-in screen, which is precisely what happens
/// on a plane, on the subway, or in a dead zone: the one place a meditation app is
/// most useful is where it stopped working.
///
/// The decision lives here, once, and is expressed in the taxonomy #676 already
/// introduced for the widget (`WidgetDataStore.SignedOutCause`) rather than a
/// second parallel one — so the app and the widget can never disagree about what
/// "signed out" means. `WidgetDataStore.shouldClearStoredSnapshot(on:)` remains
/// the single predicate for "is this cause authoritative enough to destroy local
/// state?"; this type only applies that same predicate to the app's own two
/// pieces of local state: the cached identity and the route.
///
/// **Pattern notes for #666 (web PWA).** Everything here is pure and has no
/// platform dependencies. A TypeScript mirror needs the same three moving parts:
/// (1) a classifier that treats only an HTTP 401 as proof the session is over,
/// (2) one shared predicate deciding whether a cause may destroy local state, and
/// (3) an outcome that keeps the user signed in from a cached identity whenever
/// that predicate says no and a cached identity exists.
public enum OfflineAuth {

    /// What the app should do after an attempt to load the current user failed.
    public enum Outcome: Sendable, Equatable {
        /// Authoritative: there is no usable session, or there is nothing local
        /// to fall back to. Route to sign-in; `cause` decides whether local state
        /// is cleared on the way.
        case signedOut(WidgetDataStore.SignedOutCause)
        /// No trustworthy answer — but a cached identity exists. Stay signed in
        /// and render from local state until the server can be reached again.
        case offline(WidgetDataStore.SignedOutCause)

        public var cause: WidgetDataStore.SignedOutCause {
            switch self {
            case let .signedOut(cause), let .offline(cause):
                return cause
            }
        }

        /// Whether the app is about to run from its local copy of identity/state.
        /// Drives the unobtrusive offline indicator.
        public var usesCachedIdentity: Bool {
            if case .offline = self { return true }
            return false
        }
    }

    /// Map a thrown `me()` error onto the shared sign-out taxonomy.
    ///
    /// Only an HTTP 401 counts as proof the session is over: it is the one answer
    /// that came *from the server* and was *about auth*. A response-less failure
    /// is `.unreachable` — that covers a real `URLError` and `APIError(status: 0)`,
    /// which is how `UITestAPIStore` models a dead network. Anything else the
    /// server said, and any answer we could not parse, is `.serverError`.
    ///
    /// The asymmetry is deliberate. Wrongly deciding "offline" costs a 401 on the
    /// next request and a trip to sign-in then; wrongly deciding "signed out"
    /// destroys local state and is the bug this exists to prevent. Only an
    /// authoritative signal gets to be destructive.
    public static func cause(for error: Error) -> WidgetDataStore.SignedOutCause {
        if let apiError = error as? APIError {
            if apiError.status == 401 { return .unauthorized }
            return apiError.status <= 0 ? .unreachable : .serverError
        }
        if error is URLError { return .unreachable }
        return .serverError
    }

    /// Resolve a failed `me()` into a route.
    ///
    /// `hasCachedIdentity` is the caller's answer for `CachedIdentityStore.load()`;
    /// without a local copy there is nothing to render, so even a transport failure
    /// has to land on sign-in (a first launch with no network, say).
    public static func outcome(for error: Error, hasCachedIdentity: Bool) -> Outcome {
        let cause = cause(for: error)
        // Deliberately the widget's predicate: a cause that may wipe the widget's
        // stored week is exactly a cause that may sign you out of the app.
        guard !WidgetDataStore.shouldClearStoredSnapshot(on: cause), hasCachedIdentity else {
            return .signedOut(cause)
        }
        return .offline(cause)
    }
}
